From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm-hwposion-support-recovery-from-ksm_might_need_to_copy-v4
Date: Wed, 1 Feb 2023 15:44:33 +0800

- update changelog and directly return ERR_PTR(-EHWPOISON) in
  ksm_might_need_to_copy() suggested HORIGUCHI NAOYA
- add back unlikely in unuse_pte()

Link: https://lkml.kernel.org/r/20230201074433.96641-1-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Miaohe Lin <linmiaohe@huawei.com>
Cc: Naoya Horiguchi <naoya.horiguchi@nec.com>
Cc: Tony Luck <tony.luck@intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/mm/ksm.c~mm-hwposion-support-recovery-from-ksm_might_need_to_copy-v4
+++ a/mm/ksm.c
@@ -2631,9 +2631,8 @@ struct page *ksm_might_need_to_copy(stru
 	if (new_page) {
 		if (copy_mc_user_highpage(new_page, page, address, vma)) {
 			put_page(new_page);
-			new_page = ERR_PTR(-EHWPOISON);
 			memory_failure_queue(page_to_pfn(page), 0);
-			return new_page;
+			return ERR_PTR(-EHWPOISON);
 		}
 		SetPageDirty(new_page);
 		__SetPageUptodate(new_page);
--- a/mm/swapfile.c~mm-hwposion-support-recovery-from-ksm_might_need_to_copy-v4
+++ a/mm/swapfile.c
@@ -1780,7 +1780,7 @@ static int unuse_pte(struct vm_area_stru
 		goto out;
 	}
 
-	if (hwposioned || !PageUptodate(page)) {
+	if (unlikely(hwposioned || !PageUptodate(page))) {
 		swp_entry_t swp_entry;
 
 		dec_mm_counter(vma->vm_mm, MM_SWAPENTS);
_
